fuse_loss/include/fuse_loss/loss_function.h

```diff
@@ -0,0 +1,284 @@
+// multiple M-estimators in table #1 from http://www.audentia-gestion.fr/research.microsoft/ZhangIVC-97-01.pdf (p. 24).
+//
+// In Ceres the loss functions evaluate:
+//
+//   rho(s) = 2 * \rho(sqrt(s))
+//
+// where s = r^2, being r the residual.
+//
+// This is because the output of rho(s) is mutiplied by 0.5 when the residual block cost is computed, either with or
+// without loss:
+// * cost w/o loss: https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L159
+// * cost w/  loss: https://github.com/ceres-solver/ceres-solver/blob/master/internal/ceres/residual_block.cc#L165
+//
+// Note that according to this, it looks like the Tukey loss function is incorrectly implemented in Ceres because it
+// must be multiplied by 2, so instead of dividing by 6 it should divide by 3. See:
+//
+//   https://github.com/ceres-solver/ceres-solver/blob/master/include/ceres/loss_function.h#L281-L282
+//
+// There is an easy workaround for this: combine TukeyLoss with ScaledLoss, using a scaled factor of 2.
+//
+// There is also a PR with a patch already sent to Ceres:
+//
+//   https://ceres-solver-review.googlesource.com/c/ceres-solver/+/16700
+namespace ceres
+{
+
+// Dynamic Covariance Scaling (DCS), equivalent to Switchable Constraints and similar to the Geman-McClure with the
+// tuning constant 'a'.
+//
+// The term is computed as:
+//
+//   rho(s) = a * (3 * s  - a) / (a + s)    for s >  a
+//   rho(s) = s                             for s <= a
+//
+// which gives the weight function:
+//
+//   rho'(s) = min { 1, (2 * a / (a + s))^2 }
+//
+// as described in Eq. 5.19 and 5.20 in:
+//
+//   http://www2.informatik.uni-freiburg.de/~agarwal/resources/agarwal-thesis.pdf (p. 89)
+//
+// that is equal to the square of the scaling factor in Eq. 15 in the original paper:
+//
+//   http://www2.informatik.uni-freiburg.de/~spinello/agarwalICRA13.pdf (p.3)
+//
+// which is also reproduced in Eq. 5.15 in:
+//
+//   http://www2.informatik.uni-freiburg.de/~agarwal/resources/agarwal-thesis.pdf (pp. 85-88)
+//
+// The M-estimator rho(s) equivalent to DCS is obtained from Eq. 5.18 and 5.19 by integration, giving the expression in
+// Eq. 5.20. This way we obtain a valid robust kernel, i.e. one that has positive definite weight function rho'(s) >= 0,
+// that can be used in an Iteratively Reweighted Least Squares (IRLS) problem. For more details, see:
+//
+//   http://www2.informatik.uni-freiburg.de/~agarwal/resources/agarwal-thesis.pdf (p. 89)
+//
+// The relation with the Geman-McClure loss function is explained in:
+//
+//   http://www2.informatik.uni-freiburg.de/~agarwal/resources/agarwal-thesis.pdf (pp. 90-91)
+//
+// The DCS rho(s) equation is equivalent to a scaled and translated generalized Geman-McClure for s > a (outlier region)
+// with a_GemanMcClure = sqrt(a_DCS), as shown in Eq. 5.28. With the rho(s) = 2 * rho(sqrt(s)) notation, where s = r^2
```

Comment:
```suggestion
// with a_GemanMcClure = sqrt(a_DCS), as shown in Eq. 5.28. With the rho(s) = 2 * \rho(sqrt(s)) notation, where s = r^2
```
It seems like you were distinguishing the Ceres rho(s) function from the standard rho(r) function with a \. I assumed that was intentional, to make things slightly less confusing.

Reply:
Yes. TBH I've been a little lazy at naming them properly, so I'm certainly abusing notation by just distinguishing them with the `\` prefix.

Anyway, changed.